glastopf/modules/handlers/emulators/rfi.py

```diff
@@ -21,12 +21,55 @@
 import re
 import logging
 import ssl
+import socket
+import requests
+from urlparse import urlparse
+from socket import inet_aton
+from struct import unpack
+from requests.utils import requote_uri
+
+
 
 import glastopf.sandbox.sandbox as sandbox
 from glastopf.modules.handlers import base_emulator
 
 logger = logging.getLogger(__name__)
 
+def check_ssrf(url):
+    hostname = urlparse(url).hostname
+
+    def ip2long(ip_addr):
+        return unpack("!L", inet_aton(ip_addr))[0]
+
+    def is_inner_ipaddress(ip):
+        ip = ip2long(ip)
+        return ip2long('127.0.0.0') >> 24 == ip >> 24 or \
+                ip2long('10.0.0.0') >> 24 == ip >> 24 or \
+                ip2long('172.16.0.0') >> 20 == ip >> 20 or \
+                ip2long('192.168.0.0') >> 16 == ip >> 16
+
+    try:
+    	# print re.match(r"^http(s)?://(.*?)$", url)
+        # if not re.match(r"^https?://.*/.*$", url):
+        if not re.match(r"^http(s)?://(.*?)$", url):
+            raise BaseException("url format error")
+        # print socket.getaddrinfo(hostname, 'http')
+        ip_address = socket.getaddrinfo(hostname, 'http')[0][4][0]
+        if is_inner_ipaddress(ip_address):
+            logger.info("inner ip address attack")
+            raise BaseException("inner ip address attack")
+        return True, "success"
+    except BaseException as e:
+        return False, str(e)
+    except:
+        return False, "unknow error"
+
+def safe_request_url(url, **kwargs):
+    success, errstr = check_ssrf(url)
+    if not success:
+        raise requests.exceptions.InvalidURL("SSRF Attack: %s" % (errstr,))
```

Comment:
was it actually a SSRF attack or just someone using the server from someone else to inject some code?